feat(toast): Toast implementation

benouat opened the pull request:
My final attempt at implementing [Bootstrap Toast](https://getbootstrap.com/docs/4.3/components/toasts/)

The PR basically introduce:
+ `<ngb-toast>` component. it just render bootstrap markup, with some inputs, and one single output that emit after a given delay in ms.
+ `NgbToastConfig` to override any of the default config values used to initialise above inputs. 
+ An overview page, explaining what are toast, how to use them, and also a full fledged how-to tutorial showing how to implement a global toast service inside any application.

---

**DISCLAIMER**: this PR only introduce a _declarative_ way of creating a toast. Don't look around for any `NgbToastService`, there is none! I did not implement it on purpose. If you really think it should be done, I am just waiting, go come challenge me !! 😆

peterblazejewicz commented on demo/src/app/components/toast/overview/toast-overview.component.html:
```diff
@@ -0,0 +1,94 @@
+  <ngbd-code lang="html" [code]="TOAST_INLINE_BASIC"></ngbd-code>
+  <br />
+  <p>
+    Nonetheless, with this inline technique, you must handle the toast's lifecycle yourself, i.e. it won't dissapper automagically.
```
`dissapper` > `disappear`

fbasso commented:
I'm not also 100% sure that letting the widget visibility only by the developer is a good thing. I mean, for some use cases, it can render thing more complex than expected.

For example, what about a stateless message, as something the app wants to display a notification few seconds, and let the widget 'die' by itself.

As a developer, I would expect something like:

```
<ngb-toast delay="1000" autohide="true">
  I am a simple static toast.
</ngb-toast>
```
... and that's it !

I need more time to thing about it :)

maxokorokov commented on src/toast/toast.spec.ts:
```diff
@@ -0,0 +1,125 @@
+
+    it('should emit hide output after default delay (500ms)', fakeAsync(() => {
+         const fixture = createTestComponent(`<ngb-toast header="header" (hide)="hide()">body</ngb-toast>`);
+         tick(500);
```
Maybe check that is wasn't emitted before ? Same for the next test

maxokorokov commented on src/toast/toast-config.ts:
```diff
@@ -0,0 +1,40 @@
+import {Injectable} from '@angular/core';
+
+export type AriaLiveValue = 'polite' | 'alert';
```
Don't see much utility in this type personally.

- it is used only in this file
- it makes documentation more complex

benouat replied:
I removed it, but I have to duplicate the type info for Typescript to be happy.
For example, I am not keen about `ariaLive: 'polite' | 'alert' = 'polite';` either... but without it does not compile

maxokorokov replied:
The big upside for me would be the change in this:

![Screen Shot 2019-05-28 at 10 27 22](https://user-images.githubusercontent.com/8074436/58462974-45732500-8133-11e9-9ea7-a996362946f4.png)

It would just say `type: 'polite' | 'alert'`

benouat commented on package.json:
```diff
@@ -81,7 +81,7 @@
     "@types/node": "~8.9.4",
     "@types/prismjs": "^1.9.0",
-    "bootstrap": "4.0.0",
+    "bootstrap": "4.2.1",
```
I am confused... this change is dependant on that. without `bootstrap@4.2.1` toast won't work... I really don't like to bump this version in another commit...

maxokorokov replied:
I mean we have several other PRs waiting that need BS 4.2 or 4.3. We won't bump BS in all of them I guess.

Just wanted to have a separate commit/PR with the bump of dependencies. Locally when testing/reviewing it's not a problem to bump it manually.

benouat commented on demo/src/app/components/toast/overview/toast-overview.component.html:
```diff
@@ -0,0 +1,94 @@
+    You could also create an interface to type your toast instead of using <code>any[]</code> here.
+  </ngb-alert>
+  <p>
+    Additionnally, a method to remove an existing toast from the collection is also implemented.
```
I did that one on purpose, to avoid having big chunk of snippet.

For me rule is simple: anytime you find yourself putting a comment into a snippet it means the snippet could be split, and comment should be part of main flow of documentation.

benouat commented on demo/src/app/components/toast/overview/toast-overview.component.html:
```diff
@@ -0,0 +1,94 @@
+    To achieve that, you can create a dedicated component, acting as a container to render all toasts in a convenient way.
+    For example, this container could be positionned using CSS property <code>position: static</code>.
+  </p>
+  <ngb-tabset>
```
I tried that. Single snippet is out of scope, has it is using 3 different languages. So it would be 3 snippets in a row. **much better with the tabset** trust me! 🤣

benouat commented on demo/src/app/shared/index.ts:
```diff
@@ -1,44 +1,30 @@
+import {NgbdPageHeaderComponent} from './page-wrapper/page-header.component';
+import {PageWrapper} from './page-wrapper/page-wrapper.component';
+import {SideNavComponent} from './side-nav/side-nav.component';
 
```
Will be solved by moving to latest clang, even for demo app. Right now, demo app is half formatted with prettier, half with clang, and the last **third** half is manually formatted! 🤣

benouat commented:
@peterblazejewicz @fbasso @maxokorokov thanks to all of you for your reviews!
I have addressed all your comments, and just pushed again.

@maxokorokov @fbasso 
- I removed the _weird_ part regarding `autohide` and close button.
- I added a new demo to illustrate clicking on close. (and added a disclaimer to the first 2 demos _clicking close won't do anything_)
- regarding the formatting stuff... I let it as I really want to take `5.0-rc` as an opportunity to clean up everything related to clang.

I'll open another PR to change slightly the _big_ demo. I'll update the danger toast to be `[autohide]="false"` and only disappear when click on the body (to illustrate closing without using &times; and `(hide)`)
